File: CPE/Exam/230523/23D4_UVA12455.py
"""
    類題：
    - 322. Coin Change
    - 416. Partition Equal Subset Sum
"""

T = int(input())

for _ in range(T):
    target = int(input())
    n = int(input())
    bars = list(map(int, input().split()))

    dp = [0] * (target + 1)
    dp = [False] * (target + 1)
    dp[0] = True
    for bar in bars:
        for i in range(target, bar - 1, -1):
            dp[i] |= dp[i - bar]

    print("YES" if dp[target] else "NO")

File: CPE/Exam/230523/23D4_UVA12455.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int t, target, n;
    cin >> t;
    while (t--) {
        cin >> target >> n;

        vector<int> bars(n);
        for (int i = 0; i < n; ++i) cin >> bars[i];
        vector<bool> dp(target + 1, false);

        dp[0] = true;
        for (int bar : bars) {
            for (int i = target; i >= bar; --i) {
                dp[i] = dp[i] | dp[i - bar];
            }
        }

        cout << (dp[target] ? "YES" : "NO") << endl;
    }
    return 0;
}